teps/0044-data-locality-and-pod-overhead-in-pipelines.md

```diff
@@ -697,179 +733,75 @@ finally:
-   taskRef:
-     name: just-unit-tests
-   workspaces:
-   - name: source-code
-     workspcae: source-code
-   - name: test-results
-     workspace: test-results
- - name: upload-results
-   runAfter: run-unit-tests
-   taskRef:
-     name: gcs-upload
-   params:
-   - name: location
-     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
-   workspaces:
-   - name: data
-     workspace: test-results
-```
-
-We could decide if we only support sequential execution, or support an entire DAG. Maybe even finally?
-
-An alternative to the above (which is an ["authoring time"](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability)
-solution) would be a "runtime" CRD, e.g. `TaskGroupRun` which could be passed multiple Tasks and run them together.
+The controller would need to determine how many TaskRuns are needed when the TaskGroup is first reconciled, due to
+[limitations associated with dynamically creating Tasks](#dynamically-created-tasks-in-pipelines).
+When the TaskGroup is first reconciled, it would create all TaskRuns needed, with those that are not ready to execute marked as "pending",
+and a pod with one container per TaskRun. The TaskGroup would store references to any TaskRuns created, and Task statuses would be stored on the TaskRuns.
 
 Pros:
-* Could be used to define different execution strategies
+* Creating a single TaskRun for each Task would allow individual Task statuses to be surfaced separately.
+* Allows us to choose which Pipeline features to support, and marks a clear distinction for users between supported and unsupported features.
+* Having the Pipeline controller create TaskRuns up front (as "pending" or similar) might have other benefits, for
+  example we've struggled in the past with how to represent the status of Tasks in a Pipeline which don't have a
+  backing TaskRun, e.g. they are skipped or cancelled. Now there actually would be a TaskRun backing them.
 
 Cons:
-* The line between this and a Pipeline seems very thin
+* Unclear benefit compared to adding a grouping syntax within a Pipeline and letting the PipelineRun controller handle scheduling
+* We would likely end up supporting features like `finally` for both Pipelines and TaskGroups
+(and generally reusing a lot of the PipelineRun controller's code in the TaskGroup controller)
+* Must create all TaskRuns in advance
 * New CRD to contend with
-
-### Custom scheduler
-
-In this approach we use a custom scheduler to schedule our pods.
-
-Cons:
-* [@jlpetterson has explored this option](https://docs.google.com/document/d/1lIqFP1c3apFwCPEqO0Bq9j-XCDH5XRmq8EhYV_BPe9Y/edit#heading=h.18c0pv2k7d1a)
-  and felt it added more complexity without much gain (see also https://github.com/tektoncd/pipeline/issues/3052)
-    * For example the issue [#3049](https://github.com/tektoncd/pipeline/issues/3049)
-* Doesn't help us with the overhead of multiple pods (each Task would still be a pod)
+* Extra complexity for Task/Pipeline authors
```

Comment:
also [reusability design principle](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability): "existing features should be reused when possible instead of adding new ones"